---
title: "Class 15"
author: 'Natasha ( PID: A15393874)'
date: "11/19/2021"
output: pdf_document
---
This week we are looking at differential expression analysis.
The data for this hands-on session comes from a published RNA-seq experiment where airway smooth
muscle cells were treated with dexamethasone, a synthetic glucocorticoid steroid with anti-inflammatory
effects (Himes et al. 2014).

Import/Read the data from Himes et al.
```{r}
counts <- read.csv("airway_scaledcounts.csv", row.names=1)
metadata <- read.csv("airway_metadata.csv")
```

```{r}
head(metadata)
```
```{r}
all(metadata$id == colnames(counts))
```
> Q1. How many genes are in this dataset?
There are 38694 genes in this dataset
> Q2. How many ‘control’ cell lines do we have?

```{r}
n.control <- sum(metadata$dex == "control")

```
There are 4 control cell lines in this dataset.
Extract and summarize the control samples.
To find out where the control samples are we need the metadata.

```{r}
control <- metadata[metadata$dex == "control",]
control.counts <- (counts[ , control$id])
control.mean <- rowMeans(control.counts)
head(control.mean)
```
> Q3. How would you make the above code in either approach more robust?

```{r}
control.inds <- metadata$dex == "control"
control.counts <- counts [ ,control.inds]
head(control.counts)

```
```{r}
control.mean <- rowMeans(control.counts)

```
> Q4. Follow the same procedure for the treated samples (i.e. calculate the mean per gene across drug treated samples and assign to a labeled vector called treated.mean)

```{r}
treated.inds <- metadata$dex == "treated"
treated.counts <- counts [ ,treated.inds]
head(treated.counts)

```
```{r}
treated.mean <- rowMeans(treated.counts)
```

Combine meancount data
```{r}
meancounts <- data.frame(control.mean, treated.mean)
colSums(meancounts)
```
> Q5 (a). Create a scatter plot showing the mean of the treated samples against the mean of the control samples. Your plot should look something like the following.

```{r}
plot(x=control.mean, y=treated.mean, xlab="Control", ylab="Treated")

```
> Q5 (b).You could also use the ggplot2 package to make this figure producing the plot below. What geom_?() function would you use for this plot?

```{r}
library(ggplot2)
ggplot(data=counts) +
aes (x=control.mean, y=treated.mean) +
geom_point()

```
We could use geom_point to make this graph.

> Q6. Try plotting both axes on a log scale. What is the argument to plot() that allows you to do this?

```{r}
plot(meancounts, log="xy")

```
The argument log=“xy” allows for plotting both axes on the log scale.

We often use log2 in this field because it has nice math properties that make interpretation easier.

```{r}
log2(10/10)
```
```{r}
log2(20/10)
```
```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```
Exclude columns with NaN & Inf in log2fc by excluding the genes (i.e. rows) with 0 counts. We can’t say anything about these, as we have no data for them!

```{r}
head(meancounts[,1:2] == 0)
```
Use the which() function with the ‘arr.ind=TRUE’ argument to get the columns & rows where the TRUE values (i.e the zero counts in our case) are located.

```{r}
zero.vals <- which(meancounts[,1:2] ==0, arr.ind=TRUE)
head(zero.vals)

```
> Q7. What is the purpose of the arr.ind argument in the which() function call above? Why would we then take the first column of the output and need to call the unique() function?

The purpose of the arr.ind function in our which() function all above it to return the column and row locations where the clause is true. In our problem this allows us to figure out where the rows are with 0 values. The unique () function makes sure that samples with a 0 in both the columns aren’t counted twice.

```{r}
to.rm <- unique(zero.vals[,"row"])
head(sort(to.rm))

```
Now, remove these from our meancounts dataframe.
```{r}
mycounts <- meancounts[-to.rm,]
head(mycounts)
```
How many do wwe have left?
```{r}
nrow(mycounts)

```
> Q8. Using the up.ind vector above can you determine how many up regulated genes we have at the greater than 2 fc level?

```{r}
up.ind <- mycounts$log2fc > 2
sum(up.ind ==TRUE)

```
There are 250 up regulated genes at the greater than 2 fc level.

> Q9. Using the down.ind vector above can you determine how many down regulated genes we have at the greater than 2 fc level?

```{r}
down.ind <- mycounts$log2fc < (-2)
sum(down.ind == TRUE)
```
There are 367 down regulated genes at the greater than 2 fc level.

> Q10. Do you trust these results? Why or why not? No because we have not determined whether the data is statistically significant or not. This could cause a statistically insignificant change fold.

Let’s do this the right way. DESeq2 is an R package specifically for analyzing count-based NGS data like RNA-seq. It is available from Bioconductor.

```{r}
library(DESeq2)

```
```{r}
citation("DESeq2")

```
Setup required

```{r}
dds <- DESeqDataSetFromMatrix(countData=counts,
colData=metadata,
design=~dex)
```
```{r}
dds

```
Now we can run our DESeq analysis
```{r}
dds <- DESeq(dds)
```

To get the results, we use the deseq results() function:
```{r}
res <- results(dds)
head(res)

```
```{r}
summary(res)

```
```{r}
res05 <- results(dds, alpha=0.05)
summary(res05)

```
```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

```
```{r}
#BiocManager::install("AnnotationDbi")
#BiocManager::install("org.Hs.eg.db")
```

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

```

```{r}
columns(org.Hs.eg.db)

```
```{r}
res$symbol <- mapIds(org.Hs.eg.db,
keys=row.names(res), # Our genenames
keytype="ENSEMBL", # The format of our genenames
column="SYMBOL", # The new format we want to add
multiVals="first")
```
```{r}
head(res)

```
> Q11. Run the mapIds() function two more times to add the Entrez ID and UniProt accession and GENENAME as new columns called resentrez, resuniprot and res$genename.

```{r}
res$entrez <- mapIds(org.Hs.eg.db,
keys=row.names(res),
column="ENTREZID",
keytype="ENSEMBL",
multiVals="first")
```

```{r}
res$uniprot <- mapIds(org.Hs.eg.db,
keys=row.names(res),
column="UNIPROT",
keytype="ENSEMBL",
multiVals="first")

```

```{r}
res$genename <- mapIds(org.Hs.eg.db,
keys=row.names(res),
column="GENENAME",
keytype="ENSEMBL",
multiVals="first")
```
```{r}
head(res)

```
```{r}
ord <- order( res$padj )
#View(res[ord,])
head(res[ord,])

```
```{r}
write.csv(res[ord,], "deseq_results.csv")
```

```{r}
plot( res$log2FoldChange, -log(res$padj),
xlab="Log2(FoldChange)",
ylab="-Log(P-value)")
```
```{r}
plot( res$log2FoldChange, -log(res$padj),
ylab="-Log(P-value)", xlab="Log2(FoldChange)")
# Add some cut-off lines
abline(v=c(-2,2), col="darkgray", lty=2)
abline(h=-log(0.05), col="darkgray", lty=2)
```
```{r}
# Setup our custom point color vector
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"
inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"
# Volcano plot with custom colors
plot( res$log2FoldChange, -log(res$padj),
col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)" )
# Cut-off lines
abline(v=c(-2,2), col="gray", lty=2)
abline(h=-log(0.1), col="gray", lty=2)
```
```{r}
#BiocManager::install("EnhancedVolcano")
```

```{r}
library(EnhancedVolcano)
```
```{r}
x <- as.data.frame(res)
EnhancedVolcano(x,
lab = x$symbol,
x = 'log2FoldChange',
y = 'pvalue')

```
Let’s make a commonly produced visualization from this data, namely a so-called Volcano plot. These summary figures are frequently used to highlight the proportion of genes that are both significantly regulated
and display a high fold change.

```{r}
plot(res$log2FoldChange, res$padj)
```
This is not a useful plot because all the small p-values are hidden at the bottom of the plot, so we can’t see them. Log will help.
```{r}
plot(res$log2FoldChange, log(res$padj))

```
We can filp this pvalue axis by putting a minus sign on -log(res$padj). Then, we will have a classic volcano plot.

```{r}
plot(res$log2FoldChange, -log(res$padj))
```
```{r}
# Run in your R console (i.e. not your Rmarkdown doc!)
#BiocManager::install( c("pathview", "gage", "gageData") )
```

```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)

# Examine the first 2 pathways in this kegg set for humans
head(kegg.sets.hs, 2)
```

Prep before using gage() function
```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)

```
get the results
```{r}
keggres = gage(foldchanges, gsets=kegg.sets.hs)
attributes(keggres) 
```
Look at the first three down (less) pathways
```{r}
head(keggres$less, 3)
```
use pathview() function to make a pathway plot
```{r}
pathview(gene.data=foldchanges, pathway.id="hsa05310")
# A different PDF based output of the same data
  pathview(gene.data=foldchanges, pathway.id="hsa05310", kegg.native=FALSE)
```
> Q12. Can you do the same procedure as above to
plot the pathview figures for the top 2 down-reguled pathways?
Yes

```{r}
#Graft-versus-host disease
    pathview(gene.data=foldchanges, pathway.id="hsa05332")
    pathview(gene.data=foldchanges, pathway.id="hsa05332", kegg.native=FALSE)
  #Type I diabetes mellitus
    pathview(gene.data=foldchanges, pathway.id="hsa04940")
    pathview(gene.data=foldchanges, pathway.id="hsa04940", kegg.native=FALSE)
```



